In [1]:
#importa las librerias necesarias y las funciones creadas 
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt

from src import soporte_eda as se

In [2]:
#configura para poder visualizar todas las columnas del dataframe
pd.set_option("display.max_columns", None)

In [3]:
#lee el pickle que contiene el df final 
df = pd.read_pickle("datos/datos-final.pkl")
df.head()

CÓDIGO ÓRGÃO SUPERIOR       NOME ÓRGÃO SUPERIOR  CÓDIGO ÓRGÃO  \
0                  63000  Advocacia-Geral da União         63000   
1                  63000  Advocacia-Geral da União         63000   
2                  63000  Advocacia-Geral da União         63000   
3                  63000  Advocacia-Geral da União         63000   
4                  63000  Advocacia-Geral da União         63000   

                                          NOME ÓRGÃO  CÓDIGO UNIDADE GESTORA  \
0  Advocacia-Geral da União - Unidades com víncul...                  110060   
1  Advocacia-Geral da União - Unidades com víncul...                  110060   
2  Advocacia-Geral da União - Unidades com víncul...                  110060   
3  Advocacia-Geral da União - Unidades com víncul...                  110060   
4  Advocacia-Geral da União - Unidades com víncul...                  110060   

                            NOME UNIDADE GESTORA CATEGORIA ECONÔMICA  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
3  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
4  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              ORIGEM RECEITA                                ESPÉCIE RECEITA  \
0  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
1  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   
2  Outras Receitas Correntes  Multas administrativas, contratuais e judicia   
3  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
4  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   

                                    DETALHAMENTO  VALOR PREVISTO ATUALIZADO  \
0  REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC                        0.0   
1   RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                        0.0   
2                  OUTRAS MULTAS E JUROS DE MORA                        0.0   
3      REC.DIV.ATIVA POR INFRAÇÃO ADMINISTRATIVA                        0.0   
4                            OUTRAS RESTITUICOES                        0.0   

   VALOR LANÇADO  VALOR REALIZADO  PERCENTUAL REALIZADO DATA LANÇAMENTO  \
0            0.0          1297.13                   0.0      2013-12-31   
1            0.0      26666621.42                   0.0      2013-12-31   
2            0.0        301251.13                   0.0      2013-12-31   
3            0.0          1855.58                   0.0      2013-12-31   
4            0.0         52140.68                   0.0      2013-12-31   

   ANO EXERCÍCIO  
0           2013  
1           2013  
2           2013  
3           2013  
4           2013

### Normalizar los datos

En primer lugar, los nombres de las columnas voy a dejarlos todos por igual: en mayuscula la primera letra de la primera palabra y el resto todo minúsculas. Además, en vez de espacios entre las palabras, voy a poner guión bajo

In [4]:
#saca el nombre de las columnas
df.columns

Index(['CÓDIGO ÓRGÃO SUPERIOR', 'NOME ÓRGÃO SUPERIOR', 'CÓDIGO ÓRGÃO',
       'NOME ÓRGÃO', 'CÓDIGO UNIDADE GESTORA', 'NOME UNIDADE GESTORA',
       'CATEGORIA ECONÔMICA', 'ORIGEM RECEITA', 'ESPÉCIE RECEITA',
       'DETALHAMENTO', 'VALOR PREVISTO ATUALIZADO', 'VALOR LANÇADO',
       'VALOR REALIZADO', 'PERCENTUAL REALIZADO', 'DATA LANÇAMENTO',
       'ANO EXERCÍCIO'],
      dtype='object')

In [5]:
#diccionario pasado por chatgpt con las indicaciones mias en el renombrado de las columnas 
renombrado_columnas = {
    'CÓDIGO ÓRGÃO SUPERIOR': 'Código_órgão_superior',
    'NOME ÓRGÃO SUPERIOR': 'Nome_órgão_superior',
    'CÓDIGO ÓRGÃO': 'Código_órgão',
    'NOME ÓRGÃO': 'Nome_órgão',
    'CÓDIGO UNIDADE GESTORA': 'Código_unidade_gestora',
    'NOME UNIDADE GESTORA': 'Nome_unidade_gestora',
    'CATEGORIA ECONÔMICA': 'Categoria_econômica',
    'ORIGEM RECEITA': 'Origem_receita',
    'ESPÉCIE RECEITA': 'Espécie_receita',
    'DETALHAMENTO': 'Detalhamento',
    'VALOR PREVISTO ATUALIZADO': 'Valor_previsto_atualizado',
    'VALOR LANÇADO': 'Valor_lançado',
    'VALOR REALIZADO': 'Valor_realizado',
    'PERCENTUAL REALIZADO': 'Percentual_realizado',
    'DATA LANÇAMENTO': 'Data_lançamento',
    'ANO EXERCÍCIO': 'Ano_exercício'
}


In [6]:
#renombrado de columnas
df.rename(columns=renombrado_columnas, inplace=True)
df.sample()

Código_órgão_superior     Nome_órgão_superior  Código_órgão  \
124652                  26000  Ministério da Educação         26407   

                      Nome_órgão  Código_unidade_gestora Nome_unidade_gestora  \
124652  Instituto Federal Goiano                  158124            IF GOIANO   

       Categoria_econômica             Origem_receita  \
124652  Receitas Correntes  Outras Receitas Correntes   

                                    Espécie_receita  \
124652  Indenizações, restituições e ressarcimentos   

                         Detalhamento  Valor_previsto_atualizado  \
124652  OUTRAS RESTITUICOES-PRINCIPAL                        0.0   

        Valor_lançado  Valor_realizado  Percentual_realizado Data_lançamento  \
124652            0.0          4245.16                   0.0      2016-03-31   

        Ano_exercício  
124652           2016

Ahora, por otro lado, a simple vista se observa que hay algunas columnas cuyo contenido está en mayúsculas entero, otras combinan mayúsculas con minúsculas. Para que toda la base de datos este más homogénea visualmente he decidido que voy a generar una función que vaya iterando por cada columna y deje el contenido de la siguiente forma: primera palabra en mayúscula la primera letra, el resto de las palabras todas en minúscula. Además, voy a eliminar espacios al principio y al final de cada string para asegurar que no hay ningún espacio donde no tenga que estar

In [7]:
#se genera la funcion que va a capitalizar el contenido de cada registro por cada columna
def normalizar_columnas(columna):
    return columna.str.capitalize().str.strip()

In [8]:
#se aplica la funcion a las columnas de tipo string
df[["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento"]] = df[["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento"]].apply(normalizar_columnas)
df.sample()

Código_órgão_superior     Nome_órgão_superior  Código_órgão  \
318750                  26000  Ministério da educação         26285   

                                               Nome_órgão  \
318750  Fundação universidade federal de são joão del-rei   

        Código_unidade_gestora                          Nome_unidade_gestora  \
318750                  154069  Fundacao universidade federal de s.j.del-rei   

       Categoria_econômica       Origem_receita  \
318750  Receitas correntes  Receita de serviços   

                                     Espécie_receita  \
318750  Serviços administrativos e comerciais gerais   

                                       Detalhamento  \
318750  Serv.administrat.e comerciais gerais-princ.   

        Valor_previsto_atualizado  Valor_lançado  Valor_realizado  \
318750                        0.0            0.0            175.0   

        Percentual_realizado Data_lançamento  Ano_exercício  
318750                   0.0      2017-05-16           2017

Ahora ya, a pesar de que siga en portugués, visualmente parece otra cosa

### Observación de nulos, duplicados y tipo de datos

In [9]:
#invoco a la funcion de reporte
se.info_reporte(df)

Numero_nulos  Porcentaje_nulos       Tipo_dato
Código_órgão_superior                 0              0.00           int64
Nome_órgão_superior                   0              0.00          object
Código_órgão                          0              0.00           int64
Nome_órgão                            0              0.00          object
Código_unidade_gestora                0              0.00           int64
Nome_unidade_gestora                  0              0.00          object
Categoria_econômica                   0              0.00          object
Origem_receita                        0              0.00          object
Espécie_receita                       0              0.00          object
Detalhamento                          0              0.00          object
Valor_previsto_atualizado             0              0.00         float64
Valor_lançado                         0              0.00         float64
Valor_realizado                       0              0.00         float64
Percentual_realizado                  0              0.00         float64
Data_lançamento                     578              0.06  datetime64[ns]
Ano_exercício                         0              0.00           int64

In [10]:
#comprueba si hay o no duplicados
df.duplicated().sum()

np.int64(0)

Como era de esperar por el análsis exploratorio individual ya hecho, el tipo de dato de las columnas ya es correcto, pues ya se corrigió en cada dataframe individual por año, no hay duplicados y solo hay nulos en la columna de fecha de registro, pero es un porcentaje mínimo (0,06%) y se van a mantener como están

### Eliminación de las columnas de códgios

Antes de continuar con el análisis exploratorio, hay 3 columnas las cuales creo que dan información redundante y que no son necesarias en el análisis. Las columnas Código_órgão_superior, Código_órgão y Código_unidade_gestora son 3 columnas de las que se puede prescindir, pues disponemos de la descripción de cada código en otras 3 columnas más, y son estas 3 las que se van a emplear para los análisis, pues las columnas de códigos no se interpretan bien, ni son apropiadas para posteriormente la visualización y explicación de resultados. En caso de estar trabajando con consultas de SQL estas 3 columnas podrían ser interesantes en caso de tener que realizar algún join (al fin y al cabo es más sencillo unir por códigos que por un texto específico en el que hay más probabilidades de equivocación a la hora de escribirlo), pero como no es el caso, voy a pasar a eliminarlas y a mantener las 3 (Nome_órgão_superior, Nome_órgão, Nome_unidade_gestora) que verdaderamente interesan. Antes de eliminarlas voy a comprobar que los códigos sean correctos

In [11]:
#saca el numero de codigos de organos superiores que hay
len(df["Código_órgão_superior"].unique())

25

In [12]:
#saca el numero de nombres de organos superiores que hay
len(df["Nome_órgão_superior"].unique())

25

In [13]:
#saca el numero de codigos de organos que hay
len(df["Código_órgão"].unique())

291

In [14]:
#saca el numero de nombres de organos que hay
len(df["Nome_órgão"].unique())

287

In [15]:
#saca el numero de codigos de unidades gestoras que hay
len(df["Código_unidade_gestora"].unique())

364

In [16]:
#saca el numero de nombres de unidades gestoras que hay
len(df["Nome_unidade_gestora"].unique())

356

Esta claro que algo pasa en la columna Código_órgão y Código_unidade_gestora, pues hay más códigos que nombres

In [17]:
#cuenta los codigos unicos para cada organo
unicos_nombre_organo = df.groupby("Nome_órgão")["Código_órgão"].nunique()
unicos_nombre_organo = unicos_nombre_organo.reset_index()
unicos_nombre_organo[unicos_nombre_organo["Código_órgão"] == 2]

Nome_órgão  Código_órgão
7      Agência nacional de transportes aquaviários             2
121                      Fundo nacional antidrogas             2
131                        Fundo nacional do idoso             2
133  Fundo nacional para a criança e o adolescente             2

In [18]:
#cuenta los codigos unicos para cada unidad gestora
unicos_nombre_gestora = df.groupby("Nome_unidade_gestora")["Código_unidade_gestora"].nunique()
unicos_nombre_gestora = unicos_nombre_gestora.reset_index()
unicos_nombre_gestora[unicos_nombre_gestora["Código_unidade_gestora"] == 2]

Nome_unidade_gestora  Código_unidade_gestora
8      Agencia nacional de transportes aquaviarios                       2
60       Coordenacao-geral de orcamento e financas                       2
63        Coordenacao-geral de recursos logisticos                       2
79    Diretoria de financas-sistema para o pais/mm                       2
153                      Fundo nacional antidrogas                       2
162                        Fundo nacional do idoso                       2
164  Fundo nacional para a crianca e o adolescente                       2
292    Setorial prog.orcamentaria e financeira-sfb                       2

Parece que para el nombre del órgano, tenemos cuatro órganos que disponen de dos códigos, y para el nombre de unidad gestora contamos con 8 que disponen de dos códigos. Además, el nombre de los 4 órganos coincide con 4 de los nombres de las unidades gestoras

Primer ejemplo en el que apoyo la explicación

In [19]:
#devuelve los registros donde el nombre del órgano es Fundo nacional antidrogas y el año de ejercicio 2017
ejemplo1 = df[(df["Nome_órgão"] == "Fundo nacional antidrogas") & (df["Ano_exercício"] == 2017)]
ejemplo1

Código_órgão_superior                        Nome_órgão_superior  \
352656                  30000  Ministério da justiça e segurança pública   
352657                  30000  Ministério da justiça e segurança pública   
352658                  30000  Ministério da justiça e segurança pública   
352659                  30000  Ministério da justiça e segurança pública   
352660                  30000  Ministério da justiça e segurança pública   
...                       ...                                        ...   
398561                  20000                   Presidência da república   
398562                  20000                   Presidência da república   
398563                  20000                   Presidência da república   
398564                  20000                   Presidência da república   
398565                  20000                   Presidência da república   

        Código_órgão                 Nome_órgão  Código_unidade_gestora  \
352656         30912  Fundo nacional antidrogas                  200246   
352657         30912  Fundo nacional antidrogas                  200246   
352658         30912  Fundo nacional antidrogas                  200246   
352659         30912  Fundo nacional antidrogas                  200246   
352660         30912  Fundo nacional antidrogas                  200246   
...              ...                        ...                     ...   
398561         20117  Fundo nacional antidrogas                  110246   
398562         20117  Fundo nacional antidrogas                  110246   
398563         20117  Fundo nacional antidrogas                  110246   
398564         20117  Fundo nacional antidrogas                  110246   
398565         20117  Fundo nacional antidrogas                  110246   

             Nome_unidade_gestora Categoria_econômica  \
352656  Fundo nacional antidrogas  Receitas correntes   
352657  Fundo nacional antidrogas  Receitas correntes   
352658  Fundo nacional antidrogas  Receitas correntes   
352659  Fundo nacional antidrogas  Receitas correntes   
352660  Fundo nacional antidrogas  Receitas correntes   
...                           ...                 ...   
398561  Fundo nacional antidrogas  Receitas correntes   
398562  Fundo nacional antidrogas  Receitas correntes   
398563  Fundo nacional antidrogas  Receitas correntes   
398564  Fundo nacional antidrogas  Receitas correntes   
398565  Fundo nacional antidrogas  Receitas correntes   

                   Origem_receita  \
352656  Outras receitas correntes   
352657  Outras receitas correntes   
352658  Outras receitas correntes   
352659  Outras receitas correntes   
352660  Outras receitas correntes   
...                           ...   
398561  Outras receitas correntes   
398562  Outras receitas correntes   
398563  Outras receitas correntes   
398564  Outras receitas correntes   
398565  Outras receitas correntes   

                                    Espécie_receita  \
352656                    Demais receitas correntes   
352657                    Demais receitas correntes   
352658                    Demais receitas correntes   
352659                    Demais receitas correntes   
352660                    Demais receitas correntes   
...                                             ...   
398561                    Demais receitas correntes   
398562                    Demais receitas correntes   
398563                    Demais receitas correntes   
398564                    Demais receitas correntes   
398565  Indenizações, restituições e ressarcimentos   

                                        Detalhamento  \
352656           Outras receitas-primarias-principal   
352657           Outras receitas-primarias-principal   
352658           Outras receitas-primarias-principal   
352659           Outras receitas-primarias-principal   
352660           Outras receitas-primarias-principal   
...                                              ...   
398561     

In [20]:
#comprueba que el organo Fundo nacional antidrogas solo tenga dos posibles organos superiores 
ejemplo1.groupby("Nome_órgão_superior")["Nome_órgão"].count()

Nome_órgão_superior
Ministério da justiça e segurança pública    1269
Presidência da república                       13
Name: Nome_órgão, dtype: int64

In [21]:
#comprueba que el numero de veces que aparece cada codigo del organo es igual al numero de veces que aparece cada organo superior
ejemplo1.groupby("Código_órgão")["Nome_órgão"].count()

Código_órgão
20117      13
30912    1269
Name: Nome_órgão, dtype: int64

Segundo ejemplo en el que apoyo la explicación

In [22]:
#devuelve los registros donde el nombre de la unidad gestora es Coordenacao-geral de orcamento e financas y el año de ejercicio 2016
ejemplo2 = df[(df["Nome_unidade_gestora"] == "Coordenacao-geral de orcamento e financas") & (df["Ano_exercício"] == 2016)]
ejemplo2

Código_órgão_superior                           Nome_órgão_superior  \
27795                   24000  Ministério da ciência, tecnologia, inovações   
27796                   24000  Ministério da ciência, tecnologia, inovações   
27797                   24000  Ministério da ciência, tecnologia, inovações   
27798                   24000  Ministério da ciência, tecnologia, inovações   
27799                   24000  Ministério da ciência, tecnologia, inovações   
...                       ...                                           ...   
175826                  41000                   Ministério das comunicações   
175827                  41000                   Ministério das comunicações   
175828                  41000                   Ministério das comunicações   
175829                  41000                   Ministério das comunicações   
175830                  41000                   Ministério das comunicações   

        Código_órgão                                         Nome_órgão  \
27795          24000  Ministério da ciência, tecnologia, inovações e...   
27796          24000  Ministério da ciência, tecnologia, inovações e...   
27797          24000  Ministério da ciência, tecnologia, inovações e...   
27798          24000  Ministério da ciência, tecnologia, inovações e...   
27799          24000  Ministério da ciência, tecnologia, inovações e...   
...              ...                                                ...   
175826         41000  Ministério das comunicações - unidades com vín...   
175827         41000  Ministério das comunicações - unidades com vín...   
175828         41000  Ministério das comunicações - unidades com vín...   
175829         41000  Ministério das comunicações - unidades com vín...   
175830         41000  Ministério das comunicações - unidades com vín...   

        Código_unidade_gestora                       Nome_unidade_gestora  \
27795                   240102  Coordenacao-geral de orcamento e financas   
27796                   240102  Coordenacao-geral de orcamento e financas   
27797                   240102  Coordenacao-geral de orcamento e financas   
27798                   240102  Coordenacao-geral de orcamento e financas   
27799                   240102  Coordenacao-geral de orcamento e financas   
...                        ...                                        ...   
175826                  410002  Coordenacao-geral de orcamento e financas   
175827                  410002  Coordenacao-geral de orcamento e financas   
175828                  410002  Coordenacao-geral de orcamento e financas   
175829                  410002  Coordenacao-geral de orcamento e financas   
175830                  410002  Coordenacao-geral de orcamento e financas   

       Categoria_econômica             Origem_receita  \
27795   Receitas correntes        Receita patrimonial   
27796   Receitas correntes        Receita patrimonial   
27797   Receitas correntes        Receita patrimonial   
27798   Receitas correntes        Receita patrimonial   
27799   Receitas correntes        Receita patrimonial   
...                    ...                        ...   
175826  Receitas correntes  Outras receitas correntes   
175827  Receitas correntes  Outras receitas correntes   
175828  Receitas correntes  Outras receitas correntes   
175829  Receitas correntes  Outras receitas correntes   
175830  Receitas correntes  Outras receitas correntes   

                                      Espécie_receita  \
27795   Exploração do patrimônio imobiliário do estad   
27796   Exploração do patrimônio imobiliário do estad   
27797   Exploração do patrimônio imobiliário do estad   
27798   Exploração do patrimônio imobiliário do estad   
27799   Exploração do patrimônio imobiliário do estad   
...                                               ...   
175826    Indenizações, restituições e ressarcimentos   
175827    Indenizações, restituições e ressarcimentos   
175828    Indenizações, restituiçõe

In [23]:
#comprueba la unidad gestora Coordenacao-geral de orcamento e financas solo tenga dos posibles organos superiores 
ejemplo2.groupby("Nome_órgão_superior")["Nome_unidade_gestora"].count()

Nome_órgão_superior
Ministério da ciência, tecnologia, inovações    1041
Ministério das comunicações                      252
Name: Nome_unidade_gestora, dtype: int64

In [24]:
#comprueba la unidad gestora Coordenacao-geral de orcamento e financas solo tenga dos posibles organos 
ejemplo2.groupby("Nome_órgão")["Nome_unidade_gestora"].count()

Nome_órgão
Ministério da ciência, tecnologia, inovações e comunicações - unidades com vínculo direto    1041
Ministério das comunicações - unidades com vínculo direto                                     252
Name: Nome_unidade_gestora, dtype: int64

In [25]:
#comprueba que el numero de veces que aparece cada codigo de la unidad gestora es igual al numero de veces que aparece cada organo superior y cada organo
ejemplo2.groupby("Código_unidade_gestora")["Nome_unidade_gestora"].count()

Código_unidade_gestora
240102    1041
410002     252
Name: Nome_unidade_gestora, dtype: int64

He dejado visibles dos ejemplos pero estos filtros los he hecho para cada uno de los órganos y unidades gestoras por cada año. Por lo que he observado, lo que parece que está ocurriendo de forma general es que una entidad que debería tener el mismo código siempre, tiene asociados códigos diferentes en función de si esa entidad está por debajo de una entidad u otra de nivel más superior. Es decir, en el primer ejemplo, parece que cuando el órgano Fundo nacional antidrogas pertenece al órgano superior Ministério da justiça e segurança pública, entonces el código asociado al órgano es 30912, y sin embargo, cuando pertenece al órgano superior Presidência da república, entonces el código asociado al órgano es 20117. En el segundo ejemplo, se observa algo parecido. En este caso, cuando la unidad gestora Coordenacao-geral de orcamento e financas pertenece al órgano Ministério da ciência, tecnologia, inovações e comunicações - unidades com vínculo direto y al órgano superior Ministério da ciência, tecnologia, inovações, entonces el código asociado a la unidad gestora es 240102. Sin embargo, cuando pertenece al órgano Ministério das comunicações - unidades com vínculo direto y al órgano superior Ministério das comunicações, entonces el código asociado a la unidad gestora es 410002. Esto relfeja una errata en los códigos de los órganos y de las unidades gestoras, pues independientemente de que haya órganos o unidades gestoras que, siendo lo mismo, puedan pertenecer a entidades superiores diferentes, deberían tener el mismo código, pues al final aunque un mismo órgano pueda depender de varios órganos superiores, ese órgano es el mismo y debería tener asociado el mismo código siempre. Es una errata a tomar en cuenta que se podría corregir, pero como bien ya había comentado, para este análisis las columnas de códigos no son útiles, por lo que se notifica de esta errata y continuo con lo que iba haciendo que era eliminar del dataframe las 3 columnas de códigos, manteniendo las de los nombres que además son las que están correctas

In [26]:
#elimina las columnas de codigos
df.drop(columns=["Código_órgão_superior", "Código_órgão", "Código_unidade_gestora"], inplace=True)

### Descriptivos de las variables

In [27]:
#saca los descriptivos de las variables numericas
df.describe().T

count                           mean  \
Valor_previsto_atualizado  1026299.0                27611421.049683   
Valor_lançado              1026299.0                 5242806.106963   
Valor_realizado            1026299.0                24980422.309138   
Percentual_realizado       1026299.0                      80.313812   
Data_lançamento              1025721  2018-09-20 12:23:18.569786368   
Ano_exercício              1026299.0                    2018.225555   

                                           min                  25%  \
Valor_previsto_atualizado         -214773743.0                  0.0   
Valor_lançado             -148347267246.910004                  0.0   
Valor_realizado           -156285934188.029999              260.905   
Percentual_realizado                 -114552.0                  0.0   
Data_lançamento            2013-12-31 00:00:00  2017-03-31 00:00:00   
Ano_exercício                           2013.0               2017.0   

                                           50%                  75%  \
Valor_previsto_atualizado                  0.0                  0.0   
Valor_lançado                              0.0                  0.0   
Valor_realizado                        3022.32            40853.855   
Percentual_realizado                       0.0                  0.0   
Data_lançamento            2018-08-27 00:00:00  2020-02-12 00:00:00   
Ano_exercício                           2018.0               2020.0   

                                           max                std  
Valor_previsto_atualizado      1603521711208.0  3393367175.127452  
Valor_lançado              357160677863.969971   810883695.355811  
Valor_realizado            771117711060.949951    1406892548.8405  
Percentual_realizado                72363772.0       71481.112902  
Data_lançamento            2021-12-03 00:00:00                NaN  
Ano_exercício                           2021.0           1.737966

Lo primero que me llama la atención, que ya es algo que había observado, son los ceros de las columnas Valor_previsto_atualizado, Valor_lançado, Percentual_realizado

Valor_previsto_atualizado

In [28]:
#registros en los que el el valor previsto actualizado es distinto de cero
df[df["Valor_previsto_atualizado"] != 0].shape[0] / df.shape[0] *100

1.93315982964029

Tan solo hay un 1,93% de los registros donde el valor previsto actualizado es distinto de cero 

Es bastante curioso que a pesar de que prácticamente en el 98% de los registros el valor previsto actualizado es cero, la media tome un valor grande, como lo es: 27.611.421 BRL y la mediana un valor de cero. Esto parece indicar que hay presencia de outliers en aquellos registros contados sobre los que disponemos de información en el valor previsto actualizado y que la media está sesgada y no es un valor representativo. El mínimo es de -214.773.743 BRL, un valor previsto muy bajo. Aún más llama la atención el máximo, que es de 1.603.521.711.208 BRL, un valor muy elevado. Además, la desviación estándar es muy alta, lo que es indicador de elevada dispersión en los datos

Por lo que parece, con la columna de Valor_lançado y Percentual_realizado ocurre parecido. El número de registros con un valor distinto de cero en estas columnas debe ser mínimo, y además, por el mínimo y el máximo, que son valores o extremadamente bajos, o extremadamente altos, parece que hay unos outliers que distorsionan completamente la media

Valor_lançado

In [29]:
#porcentaje de registros cuyo Valor_lançado es distinto de cero 
df[df["Valor_lançado"] != 0].shape[0] / df.shape[0] *100

0.7033038130213515

Tan solo el 0,7% de los registros tiene un valor lanzado distinto de cero 

Tiene una media de 5.242.806 BRL y una mediana de 0. La media está completamente sesgada por valores extremos. Hay un mínimo de -148.347.267.246 BRL y un máximo de 357.160.677.863 BRL. Por el valor de la desviación estándar, parece que hay elevada dispersión en los datos

Percentual_realizado

In [30]:
#porcentaje de registros cuyo Percentual_realizado es distinto de cero 
df[df["Percentual_realizado"] != 0].shape[0] / df.shape[0] *100

0.7299042481771881

Tan solo el 0,72% de los registros tienen un valor de percentual realizado distinto de cero

Tiene una media de 80% y una mediana de 0. La media está completamente sesgada por valores extremos. Hay un mínimo de -114552% y un máximo de 72363772%. Por el valor de la desviación estándar, parece que hay elevada dispersión en los datos

Valor_realizado

En cuanto al valor realizado, contamos con un mínimo de -156.285.934.188 BRL y un máximo de 771.117.711.060 BRL, un rango bastante amplio, que junto con el valor elevado de la desviación estándar (1.406.892.548 BRL) indican gran dispersión en los datos. En media, la cantidad recaudada por cada ingreso para todos los años es de 24.980.422 BRL, pero en cambio, el valor de la mediana es de 3.022 BRL, es decir, valores muy diferentes, lo que indica claramente presencia de valores extremos y que la media está sesgada y no es representativa. Por lo tanto, me quedo con la mediana, la cual indica que con el 50% de los ingresos a lo largo de los años se ha recaudado menos de 3.022 BRL y con el otro 50% más de 3.022 BRL. De la misma forma, el percentil 25 nos estaría indicando que el 25% de los ingresos tienen valores de recaudación menores a 260 BRL y el otro 75% mayores a 260 BRL. Por último, el percentil 75 indica que el 75% de los ingresos tienen valores de recaudación menores a 40.853 BRL y el otro 25% mayores a 40.853 BRL

In [31]:
#saca los descriptivos de las variables categoricas
df.describe(include="object").T

count unique  \
Nome_órgão_superior   1026299     25   
Nome_órgão            1026299    287   
Nome_unidade_gestora  1026299    356   
Categoria_econômica   1026299      5   
Origem_receita        1026299     15   
Espécie_receita       1026299     63   
Detalhamento          1026299   1885   

                                                                    top  \
Nome_órgão_superior                              Ministério da educação   
Nome_órgão            Ministério da economia - unidades com vínculo ...   
Nome_unidade_gestora            Setorial orcamentaria e financeira / me   
Categoria_econômica                                  Receitas correntes   
Origem_receita                                Outras receitas correntes   
Espécie_receita            Serviços administrativos e comerciais gerais   
Detalhamento                Serv.administrat.e comerciais gerais-princ.   

                        freq  
Nome_órgão_superior   361891  
Nome_órgão            119811  
Nome_unidade_gestora  108481  
Categoria_econômica   979596  
Origem_receita        322907  
Espécie_receita       278157  
Detalhamento          159039

Nome_órgão_superior

In [32]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Nome_órgão_superior")

count  %/Total
Nome_órgão_superior                                           
Ministério da educação                         361891    35.26
Ministério da economia                         207574    20.23
Ministério da defesa                            95490     9.30
Ministério da agricultura, pecuária e abastec   46183     4.50
Ministério da justiça e segurança pública       39980     3.90

En cuanto a la columna de Nome_órgão_superior, hay 25 órganos superiores distintos. Algo más de la mitad (55%) de los ingresos parece que son recaudados por dos óranos superiores, que son el Ministerio de educación y el Ministerio de economía, con la recaudación del 35,26% y 20,23% de los ingresos respectivamente. Es decir, estos dos órganos superiores concentran un porcentaje de recaudación significativo. En tercer lugar tendríamos al Ministerio de defensa con el 9,30% de la recaudación y luego ya al Ministerio de Agricultura, Ganadería y Abastecimiento y al Ministerio de Justicia y Seguridad Pública con el 4,50% y 3,905 de la recaudación respectivamente

Nome_órgão

In [33]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Nome_órgão")

count  %/Total
Nome_órgão                                                         
Ministério da economia - unidades com vínculo d...  119811    11.67
Fundo do exército                                    25559     2.49
Comando da marinha - fundo naval                     23734     2.31
Fundo nacional de cultura                            15955     1.55
Instituto brasileiro do meio ambiente e dos rec...   13745     1.34

En cuanto a la columna de Nome_órgão, hay 287 órganos distintos. En este caso, la recaudación no se concentra mayoritariamente en unos órganos concretos. En este caso, es el Ministerio de Economía - unidades con vínculo directo el más frecuente, con un 11,67% de la recaudación. Después ya tenemos al Fondo del Ejército, al Comando de la Marina - fondo naval, al Fondo Nacional de Cultura y al Instituto Brasileño del Medio Ambiente y de los Recursos Naturales Renovables, cuya proporción en la recaudación disminuye bastante, siendo de entre el 1-3%. Concretamente, con el 2,49%, el 2,31%, el 1,55% y el 1,34% de la recaudación respectivamente

Nome_unidade_gestora

In [34]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Nome_unidade_gestora")

count  %/Total
Nome_unidade_gestora                                         
Setorial orcamentaria e financeira / me       108481    10.57
Fundo do exercito                              25469     2.48
Diretoria de financas-sistema para o pais/mm   23532     2.29
Ibama - inst.brasileiro meio ambiente/matriz   13698     1.33
Embrapa-setorial                               13641     1.33

En cuanto a la columna de Nome_unidade_gestora, hay 356 unidades gestoras distintas. En este caso ocurre parecido al caso de los órganos, pues la recaudación no se concentra mayoritariamente en unas unidades gestoras concretas. Es la Sectorial presupuestaria y financiera / me la más frecuente, con un 10,57% de la recaudación. Después ya tenemos al Fondo del Ejército, a la Dirección de finanzas-sistema para el país/mm, al Ibama - instituto brasileño de medio ambiente/matriz y a Embrapa-sectorial, cuya proporción en la recaudación disminuye bastante, siendo de entre el 1-3%. Concretamente, con el 2,48%, el 2,29%, el 1,33% y el 1,33% de la recaudación respectivamente

Categoria_econômica

In [35]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Categoria_econômica")

count  %/Total
Categoria_econômica                                       
Receitas correntes                         979596    95.45
Receitas de capital                         30116     2.93
Receitas correntes - intra-orçamentárias    16235     1.58
Sem informação                                257     0.03
Receitas de capital - intra-orçamentárias      95     0.01

En cuanto a la colimna de Categoria_econômica, hay 5 categorías económicas diferentes, y hay una categoría que destaca con diferencia por encima del resto en términos de recaudación, pues supone el 95,45% de la recaudación total, y esta son los Ingresos corrientes. Es decir, prácticamente toda la recaudación del gobierno de Brasil entre 2013 y 2021 proviene de Ingresos corrientes. Por otro lado, los Ingresos de capital - intra-presupuestarios suponen un % de recaudación casi inexistente, pues es del 0,01%. De los Ingresos de capital y de los Ingresos corrientes - intra-presupuestarios se recauda un % también bastante mínimo (el 2,93% y el 1,58% respectivamente)

Origem_receita

In [36]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Origem_receita")

count  %/Total
Origem_receita                                              
Outras receitas correntes                    322907    31.46
Receita de serviços                          322319    31.41
Receita patrimonial                          163171    15.90
Impostos, taxas e contribuições de melhoria   78340     7.63
Contribuições                                 66781     6.51

En cuanto a la columna de Origem_receita, hay 15 fuentes específicas de ingresos diferentes. Hay 2 fuentes específicas que concentran un % de recaudación muy similar y juntas suponen más de la mitad de la recaudación. Hablamos de la fuente Otros ingresos corrientes e Ingresos por servicios, que concentran el 31,46% y el 31,41% de la recaudación respectivamente. En tercer lugar está la fuente de Ingresos patrimoniales con un % también relativamente alto, pues el 15,90% de la recaudación en estos años proviene de ingresos de esta fuente. Las fuentes Impuestos, tasas y contribuciones de mejora y Contribuciones ya concentran un % menor de recaudación (el 7,63% y el 6,51% respectivamente)

Espécie_receita

In [37]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Espécie_receita")

count  %/Total
Espécie_receita                                               
Serviços administrativos e comerciais gerais   278157    27.10
Indenizações, restituições e ressarcimentos    143577    13.99
Multas administrativas, contratuais e judicia  115276    11.23
Exploração do patrimônio imobiliário do estad   88862     8.66
Taxas                                           62336     6.07

En cuanto a la columna Espécie_receita, hay 63 tipos de ingresos diferentes. Los ingresos de tipo Servicios administrativos y comerciales generales representan más de una cuarta parte de la recaudación total con un 27,10%. Le siguen los ingresos de tipo Indemnizaciones, restituciones y reembolsos, con un 13,99%, y los de tipo Multas administrativas, contractuales y judiciales, con un 11,23%. Estas tres categorías juntas concentran más del 50% de la recaudación total entre 2013 y 2021. Otros ingresos como los de tipo Exploración del patrimonio inmobiliario del Estado y Tasas, suponen una recaudación menor (del 8,66% y 6,07% respectivamente)

Detalhamento

In [38]:
#llama a la funcion en src de soporte_eda que muestra para cada columna sus 5 valores unicos mas frecuentes con el recuento y el % sobre el total para cada uno
se.info_valores_unicos(dataframe=df, columna="Detalhamento")

count  %/Total
Detalhamento                                                  
Serv.administrat.e comerciais gerais-princ.    159039    15.50
Inscr.em concursos e proc.seletivos-principal   74638     7.27
Alugueis e arrendamentos-principal              71040     6.92
Multas previstas em legisl.especifica-princ.    44580     4.34
Restit.de despesas de exerc.anteriores-princ.   31513     3.07

En cuanto a la columna Detalhamento, contamos con 1885 detalles diferentes sobre los tipos de ingresos. El detalle de ingreso que es más frecuente es el de servicios administrativos y comerciales generales con el 15,50% de la recaudación. El resto de detalles tienen un % menor, lo que indica que en este caso la recaudación está algo más dispersa. Los ingresos con detalle Inscripciones en concursos y procesos selectivos y Alquileres y arrendamientos suponen un % de la recaudación del 7,27% y 6,92% respectivamente. Les siguen los detalles de ingresos de Multas previstas en legislación específica y Restituciones de gastos de ejercicios anteriores con unos porcentajes de 4,34% y 3,07% (respectivamente) sobre la recaudación total 

Hay que recalcar que en esta exploración de las columnas categóricas estoy hablando de el % de la recaudación recaudada por los órganos y unidades gestoras, así como el % de la recaudación de los distintos tipos de ingresos, pero siendo este % en términos de cantidad de veces que aparecen esos órganos o esos tipos de ingresos en la base de datos a lo largo de los años, lo que no quiere decir que los órganos con un mayor % sean los que más dinero recauden, o que los ingresos con mayor % sean los que mayor recaudación supongan en términos monetarios. En este caso se ha observado en términos de cantidad/ frecuencia en la recaudación. En los análisis se comprobará si esa frecuencia también supone el mayor % en términos monetarios o no

### Observación y entendimiento de la cantidad de ceros

In [39]:
#devuelve los registros donde Valor_previsto_atualizado es distinto de cero y Ano_exercício es 2016
valor_previsto_no_cero_2016 = df[(df["Valor_previsto_atualizado"] != 0) & (df["Ano_exercício"] == 2016)]
valor_previsto_no_cero_2016

Nome_órgão_superior  \
14095                        Advocacia-geral da união   
15635                    Controladoria-geral da união   
15736   Ministério da agricultura, pecuária e abastec   
15743   Ministério da agricultura, pecuária e abastec   
15846   Ministério da agricultura, pecuária e abastec   
...                                               ...   
207253                       Presidência da república   
207545                       Presidência da república   
207856                       Presidência da república   
208066                       Presidência da república   
208103                       Presidência da república   

                                               Nome_órgão  \
14095   Advocacia-geral da união - unidades com víncul...   
15635                        Controladoria-geral da união   
15736   Ministério da agricultura, pecuária e abasteci...   
15743         Empresa brasileira de pesquisa agropecuária   
15846   Ministério da agricultura, pecuária e abasteci...   
...                                                   ...   
207253                           Presidência da república   
207545                         Fundo de imprensa nacional   
207856                         Fundo de imprensa nacional   
208066     Instituto nacional de tecnologia da informação   
208103              Secretaria da micro e pequena empresa   

                                 Nome_unidade_gestora  \
14095   Coord. geral de orc. fin. e anal. cont. - agu   
15635              Diretoria de gestao interna/se/cgu   
15736       Coord.-geral de orcamento e financas-mapa   
15743                                Embrapa-setorial   
15846       Coord.-geral de orcamento e financas-mapa   
...                                               ...   
207253  Secretaria de administracao-set.orc.e fin./pr   
207545    Fundo de imprensa nacional/exec.orc.financ.   
207856    Fundo de imprensa nacional/exec.orc.financ.   
208066  Instituto nac.de tecnologia da informacao iti   
208103          Secretaria da micro e pequena empresa   

                             Categoria_econômica        Origem_receita  \
14095                         Receitas correntes   Receita de serviços   
15635                        Receitas de capital  Operações de crédito   
15736                         Receitas correntes         Contribuições   
15743                         Receitas correntes   Receita de serviços   
15846                         Receitas correntes  Receita agropecuária   
...                                          ...                   ...   
207253                        Receitas correntes   Receita patrimonial   
207545                        Receitas correntes    Receita industrial   
207856  Receitas correntes - intra-orçamentárias    Receita industrial   
208066                        Receitas correntes   Receita de serviços   
208103                        Receitas correntes   Receita de serviços   

                                     Espécie_receita  \
14095   Serviços administrativos e comerciais gerais   
15635         Operações de crédito - mercado externo   
15736                       Contribuições econômicas   
15743                                Outros serviços   
15846                           Receita agropecuária   
...                                              ...   
207253                           Valores mobiliários   
207545                            Receita industrial   
207856                            Receita industrial   
208066  Serviços administrativos e comerciais gerais   
208103  Serviços administrativos e comerciais gerais   

                                         Detalhamento  \
14095   Inscr.em concursos e proc.seletivos-principal   
15635   Op.de credito contratuais-merc.externo-princ.   
15736   Contr.s/apostas em competicoes hipicas-princ.   
15743                       Outros servicos-principal   
15846                  Receita agropecuaria-principal   
...                        

He decidido observar por ejemplo para el año 2016 aquellos registros donde el valor previsto actualizado es distinto de cero. Tan solo hay 1894 filas. Por lo que observo, parece que lo que hace esta columna es, para cada tipo de ingreso y órgano, recoger de forma acumulada el valor estimado anual que se va a recaudar

Tras lo observado, es lógico que la mediana en las columnas donde hay tanta presencia de cero diese cero y la media diese un valor mucho más elevado. Para haber logrado una media representativa de la columna valor previsto actualizado, se debería calcular la media pero excluyendo todos aquellos ceros que no representan una previsión de los ingresos 

Observo que hay registros donde se tiene dato tanto del valor previsto como del valor realizado, y otros donde solo se dispone del valor previsto y el valor realizado es cero

In [40]:
#devuelve las fechas de registro para las que el valor previsto actualizado es distinto de cero
valor_previsto_no_cero_2016["Data_lançamento"].unique()

<DatetimeArray>
['2016-01-21 00:00:00', '2016-06-08 00:00:00', '2016-01-13 00:00:00']
Length: 3, dtype: datetime64[ns]

No hay una única fecha para todos los ingresos en la que se haga esta estimación

In [41]:
#filtra por todas las columnas, desde Nome_órgão_superior hasta Detalhamento para que devuelva un ingreso concreto que se ha observado que tiene dato tanto en el valor previsto como en el realizado
df[(df["Nome_órgão_superior"] == "Presidência da república") & (df["Nome_órgão"] == "Fundo de imprensa nacional") & (df["Nome_unidade_gestora"] == "Fundo de imprensa nacional/exec.orc.financ.") & (df["Categoria_econômica"] == "Receitas correntes - intra-orçamentárias") & (df["Origem_receita"] == "Receita industrial") & (df["Espécie_receita"] == "Receita industrial") & (df["Detalhamento"] == "Receita industrial-principal") & (df["Ano_exercício"] == 2016)]

Nome_órgão_superior                  Nome_órgão  \
207607  Presidência da república  Fundo de imprensa nacional   
207608  Presidência da república  Fundo de imprensa nacional   
207609  Presidência da república  Fundo de imprensa nacional   
207610  Presidência da república  Fundo de imprensa nacional   
207611  Presidência da república  Fundo de imprensa nacional   
...                          ...                         ...   
207865  Presidência da república  Fundo de imprensa nacional   
207866  Presidência da república  Fundo de imprensa nacional   
207867  Presidência da república  Fundo de imprensa nacional   
207868  Presidência da república  Fundo de imprensa nacional   
207869  Presidência da república  Fundo de imprensa nacional   

                               Nome_unidade_gestora  \
207607  Fundo de imprensa nacional/exec.orc.financ.   
207608  Fundo de imprensa nacional/exec.orc.financ.   
207609  Fundo de imprensa nacional/exec.orc.financ.   
207610  Fundo de imprensa nacional/exec.orc.financ.   
207611  Fundo de imprensa nacional/exec.orc.financ.   
...                                             ...   
207865  Fundo de imprensa nacional/exec.orc.financ.   
207866  Fundo de imprensa nacional/exec.orc.financ.   
207867  Fundo de imprensa nacional/exec.orc.financ.   
207868  Fundo de imprensa nacional/exec.orc.financ.   
207869  Fundo de imprensa nacional/exec.orc.financ.   

                             Categoria_econômica      Origem_receita  \
207607  Receitas correntes - intra-orçamentárias  Receita industrial   
207608  Receitas correntes - intra-orçamentárias  Receita industrial   
207609  Receitas correntes - intra-orçamentárias  Receita industrial   
207610  Receitas correntes - intra-orçamentárias  Receita industrial   
207611  Receitas correntes - intra-orçamentárias  Receita industrial   
...                                          ...                 ...   
207865  Receitas correntes - intra-orçamentárias  Receita industrial   
207866  Receitas correntes - intra-orçamentárias  Receita industrial   
207867  Receitas correntes - intra-orçamentárias  Receita industrial   
207868  Receitas correntes - intra-orçamentárias  Receita industrial   
207869  Receitas correntes - intra-orçamentárias  Receita industrial   

           Espécie_receita                  Detalhamento  \
207607  Receita industrial  Receita industrial-principal   
207608  Receita industrial  Receita industrial-principal   
207609  Receita industrial  Receita industrial-principal   
207610  Receita industrial  Receita industrial-principal   
207611  Receita industrial  Receita industrial-principal   
...                    ...                           ...   
207865  Receita industrial  Receita industrial-principal   
207866  Receita industrial  Receita industrial-principal   
207867  Receita industrial  Receita industrial-principal   
207868  Receita industrial  Receita industrial-principal   
207869  Receita industrial  Receita industrial-principal   

        Valor_previsto_atualizado  Valor_lançado  Valor_realizado  \
207607                        0.0            0.0        354407.38   
207608                        0.0            0.0        580489.29   
207609                        0.0            0.0        703254.81   
207610                        0.0            0.0        147189.73   
207611                        0.0            0.0        269745.14   
...                           ...            ...              ...   
207865                        0.0            0.0       1157696.18   
207866                        0.0            0.0        388793.40   
207867                        0.0            0.0        875686.39   
207868                        0.0            0.0       1039808.06   
207869                        0.0            0.0        678889.49   

        Percentual_realizado Data_lançamento  Ano_exercício  
207607                   0.0      2016-07-13           2016  
207608                   0.0      2016-08-08

Hay 253 registros de este mismo ingreso que lo recauda el mismo órgano. Un registro contiene la estimación (anual, supongo) y una recaudación. El resto de los registros son recaudaciones de ese ingreso concreto en diferentes fechas. Es decir, hay múltiples recaudaciones reales para este ingreso registradas y tan solo hay una estimación que además coincide en la misma fecha que una de las recaudaciones. En este caso, la fecha en la que aparece la previsión es el 21 de enero. Sin embargo, observamos fechas tanto anteriores (como el 15 de enero) como posteriores (13 de julio por ejemplo) de ejecución de este ingreso. Esto creo que puede ocurrir así porque al fin y al cabo la previsión de cada ingreso está acumulada pero es una previsión actualizada, es decir, es una previsión que puede ser modificada o ajustada a lo largo del año por distintos factores (ajuste por las ejecuciones ya registradas en el año, condiciones económicas como nuevas leyes que hagan modificar la previsión). En los registros donde solo disponemos de la previsión, supongo que en este caso la fecha de registro se refiere más bien a la última fecha donde se ha actualizado esa previsión del ingreso. Si en esta misma fecha hay un valor realizado es porque en la misma fecha donde se actualizó por última vez la previsión del ingreso, se ejecutaría ese ingreso (coincidiría la fecha de previsión con una de las fechas de ejecución del ingreso). Por tanto, es coherente y puede pasar que la fecha de previsión sea posterior, o esté entre medias de las fechas de ejecución de ese ingreso

Por lo tanto, para poder analizar las discrepancias entre lo previsto y lo realmente recaudado, la columna de fecha de registro creo que es bastante irrelevante en el análisis. Si disponemos de una previsión que es el acumulado del año para cada ingreso, habrá que, anualmente, sacar la recaudación de cada ingreso también para poder comparar la diferencia entre lo que se esperó que se recaudaría y lo que realmente se recaudó. No tiene sentido comparar la previsión anual con una recaudación mensual, por lo tanto, en este caso, el análisis temporal lo veo coherente hacerlo por años, pero no por meses, y es por ello por lo que considero que la fecha de registro es algo irrelevante. La fecha de registro si puede ser útil para hacer un análisis temporal (por meses) de los ingresos recaudados, pero de las discrepancias entre lo previsto y recaudado no

In [45]:
#agrupa por las 7 primeras columnas + año y hace la suma del valor previsto y el valor realizado y el promedio del percentual
df_estimacion_realizado = df.groupby(["Nome_órgão_superior", "Nome_órgão", "Nome_unidade_gestora", "Categoria_econômica", "Origem_receita", "Espécie_receita", "Detalhamento", "Ano_exercício"]).agg({"Valor_realizado": "sum", "Valor_previsto_atualizado": "sum"})
df_estimacion_realizado = df_estimacion_realizado.reset_index()
df_estimacion_realizado

Nome_órgão_superior  \
0      Advocacia-geral da união   
1      Advocacia-geral da união   
2      Advocacia-geral da união   
3      Advocacia-geral da união   
4      Advocacia-geral da união   
...                         ...   
35805  Presidência da república   
35806  Presidência da república   
35807  Presidência da república   
35808  Presidência da república   
35809  Presidência da república   

                                              Nome_órgão  \
0      Advocacia-geral da união - unidades com víncul...   
1      Advocacia-geral da união - unidades com víncul...   
2      Advocacia-geral da união - unidades com víncul...   
3      Advocacia-geral da união - unidades com víncul...   
4      Advocacia-geral da união - unidades com víncul...   
...                                                  ...   
35805  Secretaria de políticas de promoção da igualda...   
35806  Secretaria de políticas de promoção da igualda...   
35807  Secretaria de políticas de promoção da igualda...   
35808  Secretaria de políticas de promoção da igualda...   
35809  Secretaria de políticas de promoção da igualda...   

                                Nome_unidade_gestora Categoria_econômica  \
0      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
1      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
2      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
3      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
4      Coord. geral de orc. fin. e anal. cont. - agu  Receitas correntes   
...                                              ...                 ...   
35805  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35806  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35807  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35808  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   
35809  Secretaria de polit. de prom.da iguald.racial  Receitas correntes   

                                    Origem_receita  \
0                                    Contribuições   
1      Impostos, taxas e contribuições de melhoria   
2      Impostos, taxas e contribuições de melhoria   
3      Impostos, taxas e contribuições de melhoria   
4                        Outras receitas correntes   
...                                            ...   
35805                    Outras receitas correntes   
35806                    Outras receitas correntes   
35807                    Outras receitas correntes   
35808                    Outras receitas correntes   
35809                    Outras receitas correntes   

                                     Espécie_receita  \
0                           Contribuições econômicas   
1                                              Taxas   
2                                              Taxas   
3                                              Taxas   
4      Bens, direitos e valores incorporados ao patr   
...                                              ...   
35805    Indenizações, restituições e ressarcimentos   
35806    Indenizações, restituições e ressarcimentos   
35807    Indenizações, restituições e ressarcimentos   
35808  Multas administrativas, contratuais e judicia   
35809  Multas administrativas, contratuais e judicia   

                                       Detalhamento  Ano_exercício  \
0             Condecine-divida ativa-multas e juros           2016   
1        Taxa de controle e fisc.ambiental-mul.jur.           2016   
2       Taxa de controle e fisc.ambiental-principal           2016   
3         Taxa de controle e fiscalizacao ambiental           2015   
4         Rec.div.ativa por infracao administrativa           2015   
...                                             ...            ...   
35805                     Restituicoes de convenios           2013   
35806         Restituicoes de convenios - primarias           2014   
35807

De esta forma, lo que estoy logrando es tener para cada ingreso recaudado por cada órgano, la previsión anual y la recaudación anual (la suma de todos los meses y días del año). En este punto disponemos de algunos ingresos para los que únicamente dispongo del valor de la previsión. En este caso, será porque se preveía recaudar x cantidad, y finalmente se recaudó cero. También, dispongo de algunos ingresos para los que ocurre el caso contrario, y es que se preveía recaudar 0 pero finalmente se recaudó x cantidad. Por último, dispongo de los que tienen valor distinto de cero tanto en la previsión como en la recaudación real final. En los 3 casos es interesante observar esas discrepancias entre lo que se esperaba recaudar y lo realmente recaudado

Guardo en un pickle tanto el dataframe de la unión con los cambios realizados (renombrado columnas etc) como este dataframe último que surge del group by

In [43]:
#guarda el dataframe final mejorado con los cambios nuevos
df.to_pickle("datos/datos-final-mejorado.pkl")

In [44]:
#guarda el dataframe del group by
df_estimacion_realizado.to_pickle("datos/datos-estimacion-realizado.pkl")

## Resumen EDA final
- Tipo de dato de las columnas correcto (ya fue tratado en los dataframes individuales)
- No hay duplicados 
- 0,06% de nulos en la columna de fecha de registro (ya fueron tratados en los dataframes individuales y se mantienen como están)
- Observación de errata en los códigos de los órganos y unidades gestoras. Explicación: Mismos órganos y unidades gestoras disponen de distintos códigos dependiendo de la entidad a la que pertenezcan en un nivel más superior.
- Observación de los ceros en la columna de valor previsto actualizado: Se dispone del valor previsto acumulado (anual) para cada ingreso de cada órgano.
- Variables. Debido a la alta presencia de ceros en algunas de las columnas y la explicación encontrada a los mismos, ni la media ni la mediana son realmente representativas ni aportan información relevante en el análisis
    - Valor_previsto_actualizado: Mínimo: -214773743 Máximo: 1603521711208
    - Valor_lanzado: Mínimo: -148347267246 Máximo: 357160677863
    - Percentual_realizado: Mínomo: -114552 Máximo: 72363772
    - Valor_realizado: Mínimo: -156285934188 Máximo: 771117711060 Mediana: 3022
    - Nome_órgão_superior: El Ministerio de Educación y el Ministerio de Economía aparecen en el 55% de las operaciones de recaudación.
    - Nome_órgão: En este caso está menos concentrado. El Ministerio de Economía - unidades con vínculo directo es el más frecuente (11,67%).
    - Nome_unidade_gestora: De igual forma, está menos concentrado. Es la Sectorial presupuestaria y financiera / me la más frecuente, con un 10,57%.
    - Categoria_econômica: Una categoría destaca con diferencia por encima del resto. El 95,45% de las recaudaciones provienen de Ingresos Corrientes.
    - Origem_receita: Casi el 63% de las recaudaciones provienen de Otros Ingresos Corrientes y de Ingresos por Servicios.
    - Espécie_receita: Más de una cuarta parte de la recaudación (27,10%) corresponde a Ingresos de tipo Servicios administrativos y comerciales generales.
    - Detalhamento: El detalle de ingreso que es más frecuente es el de servicios administrativos y comerciales generales con el 15,50%. 

## Limpieza de datos - Normalizar los datos, eliminación de columnas irrelevantes
- Se han renombrado las columnas y se ha homogeneizado el formato del contenido de las mismas.
- Se han eliminado las 3 columnas de códigos ya que no son útiles en este análisis. 